### ISSUE TITLE
Support eslint

### ISSUE THREAD

**Issue by ncuillery** [AlexGilleran/jsx-control-statements#7]
I appreciate this module and its various connectors. I use it well in an isomorphic app builded with gulp/webpack/babel.

The only problem I have is the missing support of the module by lint tools, here is the output with eslint enhance with the [react plugin](https://www.npmjs.com/package/eslint-plugin-react):
```
if.jsx
  74:9  error  'If' is not defined  react/jsx-no-undef

for.jsx
  51:11  error  'For' is not defined      react/jsx-no-undef
  53:19  error  "item" is not defined  no-undef
  54:23  error  "item" is not defined  no-undef
```

I can easily add an exception for `If`, `Else` and `For` but I can't skip the warning for the iterator variable (`<For each="item" ...`). 

Have you faced this problem before ? Is there an eslint plugin (existing or planned) ?

**Comment by jimfb** [comment]
One possible solution is to run eslint after you run the control statements transform; that should get rid of your unused variable warnings.  Though, I agree, it would be nice to be able to run eslint before the transform.

**Comment by AlexGilleran** [comment]
Yeah, it's an issue and as yet I don't have a good solution. What _I_ do is set `"react/jsx-no-undef": false` in `.eslintrc` and declare vars in the `render` function for all of my `each="<name>"` values. It's hacky but it shuts eslint up.

Some kind of eslint plugin would be great - hopefully it could silence these errors and also check for problems like multiple children for a control statement, forgetting `each` in a `For` statement etc. I haven't looked into it yet because I was able to work around it, but given that it's obviously not just me with this problem I'll definitely take another look at it :).

**Comment by justafish** [comment]
Here's the workaround I've been using for this in grunt. After running the jstransform task, I then run eslint on the compiled files with the babel-eslint parser and eslint-plugin-react.

+1 for having an eslint plugin which checks for all the stuff @AlexGilleran listed. Even better if it would run the babel parser so I could dump jstransform altogether :D

```
jstransform: {
      all: {
        options: {
          visitors: [ require('jsx-control-statements/jstransform').visitorList ]
        },
        files: [
          {
            expand: true,
            src: ['js/react/**/*.jsx'],
            dest: 'test/jstransform',
            ext: '.jsx'
          }
        ]
      }
    },
```

```
grunt.registerMultiTask('jstransform', 'Transpile JavaScript files', function() {
    var jsTransform = require('jstransform');
    var visitors = [];
    var done = this.async();

    var options = this.options();
    grunt.verbose.writeflags(options, 'Options');

    if (options.visitors && options.visitors.length > 0) {
      options.visitors.forEach(function (visitor) {
        if (Array.isArray(visitor)) {
          visitors = visitors.concat(visitor);
        }
        else {
          visitors.push(visitor);
        }
      });
    }

    if (this.files.length < 1) {
      grunt.verbose.warn('Destination not written because no source files were provided.');
    }

    grunt.util.async.forEachSeries(this.files, function(f, nextFileObj) {
      var destFile = f.dest;

      var files = f.src.filter(function(filepath) {
        if (!grunt.file.exists(filepath)) {
          grunt.log.warn('Source file "' + filepath + '" not found.');
          return false;
        } else {
          return true;
        }
      });

      if (files.length === 0) {
        if (f.src.length < 1) {
          grunt.log.warn('Destination not written because no source files were found.');
        }

        // No src files, go to next target. Warn would have been issued above.
        return nextFileObj();
      }

      var compiled = [];
      grunt.util.async.concatSeries(files, function(file, next) {
        grunt.log.writeln('[jstransform] Transpiling ' + file.cyan + ' --> ' + destFile.cyan);

        try {
          compiled.push(jsTransform.transform(visitors, grunt.file.read(file)).code);
        } catch (e) {
          grunt.event.emit('jstransform.error', file, e);
          grunt.fail.warn(e);
        } finally {
          next();
        }
      }, function () {
        grunt.file.write(destFile, compiled.join(grunt.util.normalizelf(grunt.util.linefeed)));
        grunt.log.writeln('[jstransform] File ' + destFile.cyan + ' created.');
        nextFileObj();
      });

    }, done);
  });
```

**Comment by nkbt** [comment]
You can also add to your `.eslintrc`
```
  "globals": {
    // jsx-control-statements
    "If": true,
    "For": true,
    // .. etc
  }
```

**Comment by justafish** [comment]
@nkbt That will still throw errors because it thinks you're using a variable that doesn't exist inside For loops

**Comment by nkbt** [comment]
That does not sound good to me. I was actually thinking that pure component way in https://github.com/romac/react-if is way better and cleaner then non-standard custom transpiling.

**Comment by jimfb** [comment]
@nkbt Firstly, your solution does not solve the problem that @justafish pointed out; you still need a way to bind variables in the scope, which is the root cause of this issue.  Supporting control statements at the component level makes the most sense only if components can bind variables (explored in https://github.com/reactjs/react-future/pull/28).  I was convinced that the transform approach was better because `If` tags can't be represented at the component level in the current implementation of React.  Even if you could solve the performance ramifications of evaluating both branches, the only way to prevent type errors from propagating is to change the element evaluation order, which would require a sprawling change to jsx and the core.  Preventing type errors is critical because the most common `if` statement is to branch on the type (eg. null check or instanceof check), which will result in fatal errors if the wrong branch is taken.

Anyway, we should probably try to stay on topic; this thread is trying to address eslint support.  Maybe open a new issue if you still aren't convinced and would like to discuss this point further.

**Comment by nkbt** [comment]
@jimfb I'm definitely not going to object a chosen way, that was obviously imho. Back to topic.

Your suggestion to run transformations before linting seems like the only working approach in the case then. Alternatively lib authors could write a plugin to support untransformed syntax. The second way would be probably smarter, because usually you want to check the code you write and not the one, generated by other tool.

Problem with this approach is that code-style will almost certainly differ and it will result in eslint giving warnings/errors anyway. We've got similar issue using babel-eslint before, since its generated code could not pass our quite strict stylistic rules (airbnb style guide), so we just stopped using any transformations before linting (and this works totally fine for us as shown here https://github.com/nkbt/esnext-quickstart).

**Comment by asaf** [comment]
I can live with declaring these component names as globals in `.eslintrc`:

```js
"globals": {
  "If": true,
  "Else": true,
}
```

**Comment by AlexGilleran** [comment]
Just noticed that someone's actually already started this:

https://github.com/vkbansal/eslint-plugin-jsx-control-statements

Haven't tried it yet but looks good :)

**Comment by vkbansal** [comment]
@AlexGilleran It is not ready yet. Although the test cases pass, but there are many false-negatives in real world usage. May be I'm missing something.

**Comment by AlexGilleran** [comment]
Oh well, good to see someone having a go anyway, nice work so far :smiley:

**Comment by vkbansal** [comment]
**Update:** I have fixed a major bug and I believe that that plugin is ready for use with `If` tags.

Rule ideas and other suggestions are welcome.

**Comment by Joe-Ying** [comment]
if i use with `extends`, eslint shows:

```
'If' is not defined                                                           react/jsx-no-undef
```

My .eslintrc:
```

{
 "extends": "eslint-config-airbnb",
 "plugins": [
   "jsx-control-statements"
 ],
 "ecmaFeatures": {
   "jsx": true
 },
 "rules": {
   "jsx-control-statements/jsx-if-condition": 1,
   "jsx-control-statements/jsx-if-single-child": 1,
   "jsx-control-statements/jsx-for-single-child": 1
 }
}
```

Any idea? 
Thanks

**Comment by vkbansal** [comment]
You need to add `If` and `For` to globals.
```
{
 "extends": "eslint-config-airbnb",
 "plugins": [
   "jsx-control-statements"
 ],
 "ecmaFeatures": {
   "jsx": true
 },
 "globals": {
   "If": true,
   "For": true
 },
 "rules": {
   "jsx-control-statements/jsx-if-condition": 1,
   "jsx-control-statements/jsx-if-single-child": 1,
   "jsx-control-statements/jsx-for-single-child": 1
 }
}
```

**Comment by texttechne** [comment]
@AlexGilleran we should reference this plugin within the docs.

**Comment by texttechne** [comment]
Fixed by @vkbansal' s plugin in version 1.0.
Docs are updated accordingly.
